ArduCopter/mode_systemid.cpp

```diff
@@ -0,0 +1,325 @@
+bool ModeSystemId::init(bool ignore_checks)
+{
+    // if landed and the mode we're switching from does not have manual throttle and the throttle stick is too high
+    if (motors->armed() && copter.ap.land_complete && !copter.flightmode->has_manual_throttle()) {
+        return false;
+    }
+
+    orig_bf_feedforward = attitude_control->get_bf_feedforward();
+    waveformTime = 0.0f;
+    magnitude_scale = 1.0f;
+    tConst = 2.0f / fStart; // Two full cycles at the starting frequency
+    systemIDState = SystemID_Testing;
+
+    gcs().send_text(MAV_SEVERITY_INFO, "SystemID Starting: axis=%d", (unsigned)systemID_axis);
+
+    AP::logger().Write(
+        "SID",
+        "TimeUS,Ax,Mag,FSt,FSp,TFin,TC,TR,TFout",
+        "s--ssssss",
+        "F--------",
+        "Qdfffffff",
+        AP_HAL::micros64(),
+        (uint8_t)systemID_axis,
+        (double)magnitude,
+        (double)fStart,
+        (double)fStop,
+        (double)tFadeIn,
+        (double)tConst,
+        (double)tRec,
+        (double)tFadeOut);
+
+    return true;
+}
+
+// systemId_run - runs the systemId controller
+// should be called at 100hz or more
+void ModeSystemId::run()
+{
+    // apply simple mode transform to pilot inputs
+    update_simple_mode();
+
+    // convert pilot input to lean angles
+    float target_roll, target_pitch;
+    get_pilot_desired_lean_angles(target_roll, target_pitch, copter.aparm.angle_max, copter.aparm.angle_max);
+
+    // get pilot's desired yaw rate
+    float target_yaw_rate = get_pilot_desired_yaw_rate(channel_yaw->get_control_in());
+
+    if (!motors->armed()) {
+        // Motors should be Stopped
+        motors->set_desired_spool_state(AP_Motors::DesiredSpoolState::SHUT_DOWN);
+    } else if (copter.ap.throttle_zero) {
+        // Attempting to Land
+        motors->set_desired_spool_state(AP_Motors::DesiredSpoolState::GROUND_IDLE);
+    } else {
+        motors->set_desired_spool_state(AP_Motors::DesiredSpoolState::THROTTLE_UNLIMITED);
+    }
+
+    switch (motors->get_spool_state()) {
+    case AP_Motors::SpoolState::SHUT_DOWN:
+        // Motors Stopped
+        attitude_control->set_yaw_target_to_current_heading();
+        attitude_control->reset_rate_controller_I_terms();
+        break;
+    case AP_Motors::SpoolState::GROUND_IDLE:
+        // Landed
+        attitude_control->set_yaw_target_to_current_heading();
+        attitude_control->reset_rate_controller_I_terms();
+        break;
+    case AP_Motors::SpoolState::THROTTLE_UNLIMITED:
+        // clear landing flag above zero throttle
+        if (!motors->limit.throttle_lower) {
+            set_land_complete(false);
+        }
+        break;
+    case AP_Motors::SpoolState::SPOOLING_UP:
+    case AP_Motors::SpoolState::SPOOLING_DOWN:
+        // do nothing
+        break;
+    }
+
+    // get pilot's desired throttle
+    float pilot_throttle_scaled = get_pilot_desired_throttle();
+
+    if((fStart > 0.0f) && (fStop > 0.0f))
```

Comment:
@lthall  or @rmackay9 What is the if statement operating on?  Just the line after it or both lines?  Leonard, what was your intention with this if statement.

Reply:
Ahh crap